xskillscore/core/probabilistic.py

```diff
@@ -820,3 +822,224 @@ def _reliability(o, f, bin_edges):
+            forecasts,
+            dichotomous_category_edges,
+            dichotomous_category_edges,
+            dim=dim,
+        )
+        fpr.append(dichotomous_contingency.false_alarm_rate())
+        tpr.append(dichotomous_contingency.hit_rate())
+    tpr = xr.concat(tpr, "probability_bin")
+    fpr = xr.concat(fpr, "probability_bin")
+
+    fpr = fpr.fillna(1.0)
+    tpr = tpr.fillna(0.0)
+
+    # pad (0,0) and (1,1)
+    fpr_pad = xr.concat(
+        [
+            xr.ones_like(fpr.isel(probability_bin=0, drop=True)),
+            fpr,
+            xr.zeros_like(fpr.isel(probability_bin=-1, drop=True)),
+        ],
+        "probability_bin",
+    )
+    tpr_pad = xr.concat(
+        [
+            xr.ones_like(tpr.isel(probability_bin=0, drop=True)),
+            tpr,
+            xr.zeros_like(tpr.isel(probability_bin=-1, drop=True)),
+        ],
+        "probability_bin",
+    )
+
+    # https://github.com/scikit-learn/scikit-learn/blob/42aff4e2edd8e8887478f6ff1628f27de97be6a3/sklearn/metrics/_ranking.py#L916
+    # Attempt to drop thresholds corresponding to points in between and
+    # collinear with other points. These are always suboptimal and do not
+    # appear on a plotted ROC curve (and thus do not affect the AUC).
+    # Here xr.diff(_, 2) is used as a "second derivative" to tell if there
+    # is a corner at the point. (...) This keeps all cases where the point should be
+    # kept, but does not drop more complicated cases like fps = [1, 3, 7],
+    # tps = [1, 2, 4]; there is no harm in keeping too many thresholds.
+    if drop_intermediate and fpr.probability_bin.size > 2:
+
+        def _drop_intermediate(fpr, tpr):
+            optimal_idxs = xr.concat(
+                [
+                    fpr.isel(probability_bin=0, drop=True).astype("bool"),
+                    np.logical_or(
+                        fpr.diff("probability_bin", 2), tpr.diff("probability_bin", 2)
+                    ),
+                    fpr.isel(probability_bin=0, drop=True).astype("bool"),
+                ],
+                "probability_bin",
+            )
+            optimal_idxs["probability_bin"] = np.arange(
+                optimal_idxs.probability_bin.size
+            )
+            optimal_idxs = optimal_idxs.where(optimal_idxs, drop=True).probability_bin
+            tpr = tpr.sel(probability_bin=optimal_idxs)
+            fpr = fpr.sel(probability_bin=optimal_idxs)
+            return fpr, tpr
+
+        fpr, tpr = _drop_intermediate(fpr, tpr)
+        fpr_pad, tpr_pad = _drop_intermediate(fpr_pad, tpr_pad)
+
+    def auc(tpr, fpr, dim="probability_bin"):
+        """Get area under the curve with trapez method."""
+        area = xr.apply_ufunc(
+            np.trapz, tpr, fpr, input_core_dims=[[dim], [dim]], dask="allowed"
+        )
+        area = np.clip(area, 0, 1)  # allow only values between 0 and 1
+        return area
+
+    area = auc(-tpr_pad, fpr_pad)
+    if continuous:
+        area = 1 - area  # dirty fix
```

Comment:
This I am unsure about. Looking at the numbers /results/tests the current code works.

Reply:
wonder if `direction` helps here from sklean https://github.com/scikit-learn/scikit-learn/blob/662cc647d93117d7a76b51fc5778147bc49cbf29/sklearn/metrics/_ranking.py#L100

Reply:
Thanks for the hint. Got it sorted out.

Reply:
I tried this direction this. but our outputs are not always sorted in `tpr` which breaks this.